// notificationapi-node-server-sdk/lib/cjs/interfaces.js
"use strict";
exports.__esModule = true;
exports.PushProviders = exports.Region = exports.Channels = void 0;
/** Supporting notification Channels.*/
var Channels;
(function (Channels) {
    Channels["EMAIL"] = "EMAIL";
    Channels["INAPP_WEB"] = "INAPP_WEB";
    Channels["SMS"] = "SMS";
    Channels["CALL"] = "CALL";
    Channels["PUSH"] = "PUSH";
    Channels["WEB_PUSH"] = "WEB_PUSH";
})(Channels = exports.Channels || (exports.Channels = {}));
/** Region enum defines the available API endpoints for different geographical regions:
 * - US_REGION: United States region endpoint (https://api.notificationapi.com)
 * - EU_REGION: European Union region endpoint (https://api.eu.notificationapi.com)
 * - CA_REGION: Canada region endpoint (https://api.ca.notificationapi.com)
 */
var Region;
(function (Region) {
    Region["US_REGION"] = "https://api.notificationapi.com";
    Region["EU_REGION"] = "https://api.eu.notificationapi.com";
    Region["CA_REGION"] = "https://api.ca.notificationapi.com";
})(Region = exports.Region || (exports.Region = {}));
var PushProviders;
(function (PushProviders) {
    /**firebase-fcm token provider */
    PushProviders["FCM"] = "FCM";
    /**APN token provider */
    PushProviders["APN"] = "APN";
})(PushProviders = exports.PushProviders || (exports.PushProviders = {}));
